traffic_ops/traffic_ops_golang/server/put_status.go

```diff
@@ -24,69 +24,94 @@ import (
+	serverInfo, exists, err := dbhelpers.GetServerInfo(id, tx)
 	if err != nil {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, err)
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
 		return
 	}
 	if !exists {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusNotFound, fmt.Errorf("server ID %d not found", inf.IntParams["id"]), nil)
+		api.HandleErr(w, r, tx, http.StatusNotFound, fmt.Errorf("server ID %d not found", id), nil)
 		return
 	}
 
 	status := tc.StatusNullable{}
 	statusExists := false
 	if reqObj.Status.Name != nil {
-		status, statusExists, err = dbhelpers.GetStatusByName(*reqObj.Status.Name, inf.Tx.Tx)
+		status, statusExists, err = dbhelpers.GetStatusByName(*reqObj.Status.Name, tx)
 	} else if reqObj.Status.ID != nil {
-		status, statusExists, err = dbhelpers.GetStatusByID(*reqObj.Status.ID, inf.Tx.Tx)
+		status, statusExists, err = dbhelpers.GetStatusByID(*reqObj.Status.ID, tx)
 	} else {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("status is required"), nil)
+		api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New("status is required"), nil)
 		return
 	}
 	if err != nil {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, err)
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
 		return
 	}
 	if !statusExists {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("invalid status (does not exist)"), nil)
+		api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New("invalid status (does not exist)"), nil)
 		return
 	}
 
 	if *status.Name == tc.CacheStatusAdminDown.String() || *status.Name == tc.CacheStatusOffline.String() {
 		if reqObj.OfflineReason == nil {
-			api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("offlineReason is required for "+tc.CacheStatusAdminDown.String()+" or "+tc.CacheStatusOffline.String()+" status"), nil)
+			api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New("offlineReason is required for "+tc.CacheStatusAdminDown.String()+" or "+tc.CacheStatusOffline.String()+" status"), nil)
 			return
 		}
 		*reqObj.OfflineReason = inf.User.UserName + ": " + *reqObj.OfflineReason
 	} else {
 		reqObj.OfflineReason = nil
 	}
-	if err := updateServerStatusAndOfflineReason(inf.IntParams["id"], *status.ID, reqObj.OfflineReason, inf.Tx.Tx); err != nil {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, err)
+
+	existingStatus, existingStatusUpdatedTime := checkExistingStatusInfo(id, tx)
+	if *status.Name != "ONLINE" && *status.Name != "REPORTED" && *status.ID != existingStatus {
+		dsIDs, err := getDeliveryServicesThatOnlyHaveThisServerAssigned(id, tx)
+		if err != nil {
+			sysErr = fmt.Errorf("getting Delivery Services to which server #%d is assigned that have no other servers: %v", id, err)
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, sysErr)
+			return
+		}
+		if len(dsIDs) > 0 {
+			alerts := make([]tc.Alert, 0, len(dsIDs))
```

Comment:
I guess there's not really a good reason. Plus I'm not sure TP can handle that right now.

Reply:
Yeah, that was my other concern. I kind of imagined getting 100 alerts back and having to scroll down past 100 red alert panes (not that I know it does that or not).

Reply:
I think it might even overlay them in the same place on the page. Or maybe I'm thinking of the implementation in `experimental/traffic-portal`.

Anyway, I'll change it to one alert.